File: moviejava/app/src/main/java/com/example/moviejava/ui/home/fragment/genres/GenresViewModel.java
package com.example.moviejava.ui.home.fragment.genres;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.moviejava.data.remote.genres.GenresRepository;
import com.example.moviejava.model.Genres;
import com.example.moviejava.model.Results;
import com.example.moviejava.model.State;
import com.example.moviejava.ui.base.BaseViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class GenresViewModel extends BaseViewModel {

    private GenresRepository repository;

    private final MutableLiveData<State<List<Genres>>> _genres = new MutableLiveData<>();

    public LiveData<State<List<Genres>>> getGenres() {
        return _genres;
    }

    private final ArrayList<Results> list = new ArrayList<>();
    private final MutableLiveData<State<List<Results>>> _results = new MutableLiveData<>();

    public LiveData<State<List<Results>>> getResults() {
        return _results;
    }

    private final MutableLiveData<State<Boolean>> _isChose = new MutableLiveData<>();

    public LiveData<State<Boolean>> getChose() {
        return _isChose;
    }

    public void clearData() {
        list.clear();
    }

    @Inject
    public GenresViewModel(GenresRepository repository) {
        this.repository = repository;
        setGenres();
    }

    public void setGenres() {
        launchRxSingle(repository.getGenres(), result -> _genres.postValue(State.success(result.genres)), null);

    }

    public void setResults(int id, int page) {
        launchRxSingle(repository.getListGenresByFilm(id, page), result -> {
            list.addAll(result.results);
            _results.postValue(State.success(list));
            _isChose.postValue(State.success(true));
        }, null);
    }

}

File: movie/app/src/main/java/com/example/movie/ui/home/fragment/genres/GenresViewModel.kt
package com.example.movie.ui.home.fragment.genres

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.movie.data.remote.genres.GenresRepository
import com.example.movie.model.Genres
import com.example.movie.model.Results
import com.example.movie.model.State
import com.example.movie.ui.base.BaseViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class GenresViewModel @Inject constructor(val repository: GenresRepository) : BaseViewModel() {
    private val _genres = MutableLiveData<State<List<Genres>>>()
    val listGenres: LiveData<State<List<Genres>>>
        get() = _genres
    private val list = mutableListOf<Results>()
    private val _results = MutableLiveData<State<List<Results>>>()
    val results: LiveData<State<List<Results>>>
        get() = _results
    private val _isChose = MutableLiveData(false)
    val isChose: LiveData<Boolean>
        get() = _isChose

    fun clearData() {
        list.clear()
    }

    init {
        launchViewModel(repository.getGenres(), { _genres.postValue(State.success(it.genres)) }, {
            Log.e("err", it.message.toString())
        })

    }

    fun getListFilmByGenres(id: Int, page: Int) {
        _results.postValue(State.loading())
        launchViewModel(repository.getListFilmByGenres(id = id, page = page), {
            list.addAll(it.results)
            _results.postValue(State.success(list))
            _isChose.value = true
        }, {
            Log.e("err", it.message.toString())
        })
    }
}